solidity/contracts/libs/isms/MultisigIsmMetadata.sol

```diff
@@ -3,27 +3,26 @@ pragma solidity >=0.8.0;
 
 /**
  * Format of metadata:
- * [   0:  32] Merkle root
- * [  32:  36] Root index
- * [  36:  68] Origin mailbox address
- * [  68:1092] Merkle proof
- * [1092:????] Validator signatures, 65 bytes each, length == Threshold
+ * [   0:   4] Root index
+ * [   4:  36] Origin mailbox address
+ * [  36:  37] Merkle type
+ * [  36:  VS] Validator signatures (where VS = 36 + Threshold * 65 bytes)
+ *
+ * [  VS:????] Merkle root
+ *                OR
+ * [  VS:????] (Message ID, Merkle proof)
  */
 library MultisigIsmMetadata {
-    uint256 private constant MERKLE_ROOT_OFFSET = 0;
-    uint256 private constant MERKLE_INDEX_OFFSET = 32;
-    uint256 private constant ORIGIN_MAILBOX_OFFSET = 36;
-    uint256 private constant MERKLE_PROOF_OFFSET = 68;
-    uint256 private constant SIGNATURES_OFFSET = 1092;
-    uint256 private constant SIGNATURE_LENGTH = 65;
+    uint8 private constant MERKLE_INDEX_OFFSET = 0;
+    uint8 private constant ORIGIN_MAILBOX_OFFSET = 4;
+    uint8 private constant MERKLE_TYPE_OFFSET = 36;
+    uint8 private constant SIGNATURES_OFFSET = 37;
+    uint8 private constant SIGNATURE_LENGTH = 65;
+    uint16 private constant PROOF_LENGTH = 32 * 32;
 
-    /**
-     * @notice Returns the merkle root of the signed checkpoint.
-     * @param _metadata ABI encoded Multisig ISM metadata.
-     * @return Merkle root of the signed checkpoint
-     */
-    function root(bytes calldata _metadata) internal pure returns (bytes32) {
-        return bytes32(_metadata[MERKLE_ROOT_OFFSET:MERKLE_INDEX_OFFSET]);
+    enum MerkleType {
+        ROOT,
```

Comment:
nit: Consider `MetadataType`, `SuffixType`, or st

MerkleType implies to me that there are different types of merkle trees